pkg/utils/transport/pool.go

```diff
@@ -0,0 +1,94 @@
+package transport
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/containerd/nydus-snapshotter/pkg/utils/registry"
+	"github.com/golang/groupcache/lru"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/pkg/errors"
+)
+
+var _ Resolve = &Pool{}
+
+const HTTPClientTimeOut = time.Second * 60
+
+type Pool struct {
+	trPoolMu  sync.Mutex
+	trPool    *lru.Cache
+	transport http.RoundTripper
+}
+
+func NewPool() *Pool {
+	pool := Pool{
+		transport: http.DefaultTransport,
+		trPool:    lru.New(3000),
+	}
+	return &pool
+}
+
+type Resolve interface {
+	Resolve(ref name.Reference, digest string, keychain authn.Keychain) (string, http.RoundTripper, error)
+}
+
+func (r *Pool) Resolve(ref name.Reference, digest string, keychain authn.Keychain) (string, http.RoundTripper, error) {
+	r.trPoolMu.Lock()
+	defer r.trPoolMu.Unlock()
+	endpointURL := fmt.Sprintf("%s://%s/v2/%s/blobs/%s",
+		ref.Context().Registry.Scheme(),
+		ref.Context().RegistryStr(),
+		ref.Context().RepositoryStr(),
+		digest)
+
+	if tr, ok := r.trPool.Get(ref.Name()); ok {
+		if url, err := redirect(endpointURL, tr.(http.RoundTripper)); err != nil {
+			return url, tr.(http.RoundTripper), nil
+		}
+	}
+	r.trPool.Remove(ref.Name())
+	tr, err := registry.AuthnTransport(ref, r.transport, keychain)
+	if err != nil {
+		return "", nil, err
+	}
+	url, err := redirect(endpointURL, tr)
+	if err != nil {
+		return "", nil, err
+	}
+	r.trPool.Add(ref.Name(), tr)
+	return url, tr, nil
+}
+
+func redirect(endpointURL string, tr http.RoundTripper) (url string, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), HTTPClientTimeOut)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, "GET", endpointURL, nil)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to request to the registry")
+	}
+	req.Close = false
+	req.Header.Set("Range", "bytes=0-0")
+	res, err := tr.RoundTrip(req)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to request to %q", endpointURL)
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, res.Body)
+		res.Body.Close()
+	}()
+
+	if res.StatusCode/100 == 2 {
+		url = endpointURL
+	} else if redir := res.Header.Get("Location"); redir != "" && res.StatusCode/100 == 3 {
+		url = redir
+	} else {
+		return "", fmt.Errorf("failed to access to %q with code %v", endpointURL, res.StatusCode)
```

Comment:
In general, the response body includes some detailed error message, maybe we need to parse as text and print it out.